### ISSUE TITLE
Marshalling of byte[] arrays is causing excessive GC load (and gen2 collections)

### ISSUE THREAD

**Issue by hans-olav** [dokan-dev/dokan-dotnet#202]
My app was serving a large file (~2 GB) through Dokan and I was noticing a gen2 garbage collection every second. This is obviously not good for performance - I was seeing 30% of CPU spent in GC. After investigating, it turns out the allocations were coming from marshaling of the byte[] array in ReadFile(). A managed array was being allocated for every ReadFile() call and thrown away after the call completes.

Today's code looks like this:
```
public delegate NtStatus ReadFileDelegate(
    [MarshalAs(UnmanagedType.LPWStr)] string rawFileName,
    [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 2), Out] byte[] rawBuffer,
    uint rawBufferLength,
...
);
```

By modifying it to not marshal at all and just pass the `IntPtr` I was able to completely eliminate the GC load. The implementor of `IDokanOperations` can then use `Marshal.Copy` or unsafe code to fill the output.

This is obviously a breaking interface change, but maybe it's worth adding a new interface, something like `IDokanOperationsUnsafe`, for those that want the perf?

**Comment by Liryna** [comment]
Hi @hans-olav ,

I see the situation here. Does calling GC Collect to force the previous buffer to be release improve the situation ?

**Comment by hans-olav** [comment]
I haven't tried but I don't think it would help. The file buffers for ReadFile() are usually 1 MB for large files, which means they get allocated directly on the large object heap (LOH) instead of on the gen0 heap - the threshold for that is 85K. Only full GC (gen2 + LOH) collections collect the LOH, and those full GC runs are expensive. In my case, I was serving ~300 MB/s through Dokan, which means marshalling was allocating and throwing away 300 MB of memory every second. This is why GC was kicking in so often and burning 30% CPU.

In other words, the problem isn't that GC isn't happening - it's that full GC is _needed_ that often to keep the high allocation rate from consuming all memory. Calling GC.Collect shouldn't change that - lowering the allocation rate would.

**Comment by Liryna** [comment]
@hans-olav that’s a great answer !
Also you changes would not be an API break if the marshal is done in ReadFileProxy for the user.

Would it be possible for you to make a pull request with the changes ? This would be highly appreciated !

**Comment by hans-olav** [comment]
For the most effective solution, we would need to modify the user interface (`IDokanOperations`) as well, so we should discuss how we would want to do this. There is two levels to this optimization:

1. We can move the marshalling to `ReadFileProxy `and pool the managed array buffers within DokanDotNet. This would reduce the allocation frequency, but it would still do a extra copy of the entire buffer to marshall it. It would not require interface changes.

2. We don't marshal the buffer at all and pass the IntPtr directly through to `IDokanOperations`. With this solution there is no extra copy and better perf. The interface of `IDokanOperations.ReadFile` changes from `ReadFile(..., byte[] buffer, ...)` to `ReadFile(..., IntPtr buffer, uint bufferLength, ...)`. This is what I do in my implementation, and I use P/Invoke to `kernel32!CopyMemory` to full the buffer.

#2 is obviously a breaking change, so it probably would require a new interface so folks can choose.

What are your thoughts?

**Comment by hans-olav** [comment]
Re #1 above - the simplest implementation would be to use `ArrayPool<T>` from [System.Buffers](https://www.nuget.org/packages/System.Buffers). But it depends on .NET 4.5, which would remove .NET 4.0 as an option for Dokan. Alternatively, we could roll our own to keep compatibility with .NET 4.0.

Or, we could not pool them and make the new interface which exposes the `IntPtr `the only option to avoid GC. What do you think?

**Comment by hans-olav** [comment]
Hi,

Yeah, I'll work on a PR for it.

I have an idea that lets us do both #1 and #2: We keep the simple (today's) interface for folks that want simplicity and pool the buffers behind the scenes automatically. This way all of today's apps get the improvement. We then add a derived interface that exposes the IntPtr that users optionally implement of they want the raw perf.

Re: ArrayPool: I think we could just roll our own simple pool to retain 4.0 compat - it's not much code. Will work on that.

**Comment by hans-olav** [comment]
See PR #206 for the implementation of this feature.

**Comment by hans-olav** [comment]
#206 merged to master, closing issue.
